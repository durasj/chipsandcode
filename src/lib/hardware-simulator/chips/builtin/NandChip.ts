import type { ChipPin } from '../Chip';
import type Chip from '../Chip';

/**
 * NAND gate implemented natively
 */
class NandChip implements Chip {
  public readonly name = 'Nand' as string;

  private a: boolean[];
  private b: boolean[];
  private out: boolean[];

  constructor(private width = 1) {
    this.a = new Array(width).fill(false);
    this.b = new Array(width).fill(false);
    this.out = new Array(width).fill(false);
  }

  setInput(name: 'a' | 'b', value: boolean[]) {
    this[name] = value;
  }

  getOutput() {
    return this.out;
  }

  getPins(): Map<string, ChipPin> {
    // Constructs map on the fly
    // Pin inspection is less performance sensitive than actual execution

    return new Map([
      ['a', { type: 'input', width: this.width, state: this.a, connections: [] }],
      ['b', { type: 'input', width: this.width, state: this.b, connections: [] }],
      ['out', { type: 'output', width: this.width, state: this.out, connections: [] }],
    ]);
  }

  public run() {
    this.out = this.a.map((a, i) => !(a && this.b[i]));
  }
}

export default NandChip;
