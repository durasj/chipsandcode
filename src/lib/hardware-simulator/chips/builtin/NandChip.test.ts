import { describe, expect, it } from 'vitest';

import NandChip from './NandChip';

describe('Nand Chip', () => {
  it('Provides interface info', async () => {
    const chip = new NandChip();

    expect(chip.name).toBe('Nand');
    expect(chip.getPins()).toMatchSnapshot();
  });

  it('Has correct logic', async () => {
    const chip = new NandChip();

    const values = [
      [[false], [false], [true]],
      [[true], [false], [true]],
      [[false], [true], [true]],
      [[true], [true], [false]],
    ];

    values.forEach(([a, b, out]) => {
      chip.setInput('a', a);
      chip.setInput('b', b);
      chip.run();
      expect(chip.getOutput()).toStrictEqual(out);
    });
  });
});
