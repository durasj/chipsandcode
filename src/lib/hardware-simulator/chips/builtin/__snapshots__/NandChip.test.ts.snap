// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Nand Chip > Provides interface info 1`] = `
Map {
  "a" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "b" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "out" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "output",
    "width": 1,
  },
}
`;
